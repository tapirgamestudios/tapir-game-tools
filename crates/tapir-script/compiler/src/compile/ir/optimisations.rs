use std::{
    collections::HashMap,
    ops::{BitOr, BitOrAssign},
};

use petgraph::visit::Dfs;

use crate::{ast::SymbolId, compile::ir::TapIrFunction};

mod empty_phi;

#[derive(Clone, Copy, PartialEq, Eq)]
enum OptimisationResult {
    DidSomething,
    DidNothing,
}

impl BitOrAssign for OptimisationResult {
    fn bitor_assign(&mut self, rhs: Self) {
        *self = *self | rhs;
    }
}

impl BitOr for OptimisationResult {
    type Output = OptimisationResult;

    fn bitor(self, rhs: Self) -> Self::Output {
        if self == OptimisationResult::DidSomething || rhs == OptimisationResult::DidSomething {
            OptimisationResult::DidSomething
        } else {
            OptimisationResult::DidNothing
        }
    }
}

fn rename_all_variables(
    function: &mut TapIrFunction,
    renames: &HashMap<SymbolId, SymbolId>,
) -> OptimisationResult {
    if renames.is_empty() {
        return OptimisationResult::DidNothing;
    }

    let mut did_something = OptimisationResult::DidNothing;

    let mut dfs = Dfs::new(&*function, function.root);
    while let Some(block_id) = dfs.next(&*function) {
        let block = function.block_mut(block_id).expect("Could not find block");

        for symbol in block.sources_mut() {
            if let Some(renamed_symbol) = renames.get(symbol) {
                *symbol = *renamed_symbol;
                did_something = OptimisationResult::DidSomething;
            }
        }

        for symbol in block.targets_mut() {
            if let Some(renamed_symbol) = renames.get(symbol) {
                *symbol = *renamed_symbol;
                did_something = OptimisationResult::DidSomething;
            }
        }
    }

    did_something
}
