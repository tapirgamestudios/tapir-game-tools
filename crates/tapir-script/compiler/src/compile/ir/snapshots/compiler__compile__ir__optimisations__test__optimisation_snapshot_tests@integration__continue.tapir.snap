---
source: crates/tapir-script/compiler/src/compile/ir/optimisations.rs
expression: output
input_file: crates/tapir-script/compiler/src/compile/ir/snapshot_tests/optimisations/integration/continue.tapir
---
----------- before -------------
=======================
fn @toplevel() -> 

---- block 2 ----
i.15 = int 0
jmp 0

---- block 0 ----
i.16 = φ(i.15 from 2, i.19 from 3, i.19 from 9)
temp.17 = i.16
temp.18 = int 1
i.19 = temp.17 + temp.18
temp.20 = i.19
temp.21 = int 2
temp.22 = temp.20 %% temp.21
temp.23 = int 0
temp.24 = temp.22 == temp.23
conditional_jump temp.24 3 4

---- block 3 ----
jmp 0

---- block 4 ----
jmp 5

---- block 5 ----
getprop temp.25, 0
temp.26 = int 1
temp.27 = temp.25 + temp.26
storeprop temp.27, 0
temp.28 = i.19
temp.29 = int 20
temp.30 = temp.28 > temp.29
conditional_jump temp.30 7 8

---- block 7 ----
jmp 1

---- block 8 ----
jmp 9

---- block 1 ----
return 

---- block 9 ----
jmp 0


----------- optimised -------------
=======================
fn @toplevel() -> 

---- block 2 ----
i.15 = int 0
jmp 0

---- block 0 ----
i.16 = φ(i.15 from 2, i.19 from 0, i.19 from 5)
temp.17 = i.16
temp.18 = int 1
i.19 = i.16 + temp.18
temp.20 = i.19
temp.21 = int 2
temp.22 = i.19 %% temp.21
temp.23 = int 0
temp.24 = temp.22 == temp.23
conditional_jump temp.24 0 5

---- block 5 ----
getprop temp.25, 0
temp.26 = int 1
temp.27 = temp.25 + temp.26
storeprop temp.27, 0
temp.28 = i.19
temp.29 = int 20
temp.30 = i.19 > temp.29
conditional_jump temp.30 1 0

---- block 1 ----
return
